libraries/src/AWS.Lambda.Powertools.Logging/Internal/PowertoolsLogger.cs

```diff
@@ -228,24 +251,120 @@ public bool IsEnabled(LogLevel logLevel)
         }
 
-        message.TryAdd(LoggingConstants.KeyTimestamp, DateTime.UtcNow.ToString("o"));
-        message.TryAdd(LoggingConstants.KeyLogLevel, logLevel.ToString());
-        message.TryAdd(LoggingConstants.KeyService, Service);
-        message.TryAdd(LoggingConstants.KeyLoggerName, _name);
-        message.TryAdd(LoggingConstants.KeyMessage,
-            CustomFormatter(state, exception, out var customMessage) && customMessage is not null
-                ? customMessage
-                : formatter(state, exception));
+        logEntry.TryAdd(LoggingConstants.KeyTimestamp, timestamp.ToString("o"));
+        logEntry.TryAdd(LoggingConstants.KeyLogLevel, logLevel.ToString());
+        logEntry.TryAdd(LoggingConstants.KeyService, Service);
+        logEntry.TryAdd(LoggingConstants.KeyLoggerName, _name);
+        logEntry.TryAdd(LoggingConstants.KeyMessage, message);
+
         if (CurrentConfig.SamplingRate.HasValue)
-            message.TryAdd(LoggingConstants.KeySamplingRate, CurrentConfig.SamplingRate.Value);
+            logEntry.TryAdd(LoggingConstants.KeySamplingRate, CurrentConfig.SamplingRate.Value);
         if (exception != null)
-            message.TryAdd(LoggingConstants.KeyException, exception);
+            logEntry.TryAdd(LoggingConstants.KeyException, exception);
+
+        return logEntry;
+    }
+
+    /// <summary>
+    ///     Gets a formatted log entry.
+    /// </summary>
+    /// <param name="logLevel">Entry will be written on this level.</param>
+    /// <param name="timestamp">Entry timestamp.</param>
+    /// <param name="message">The message to be written. Can be also an object.</param>
+    /// <param name="exception">The exception related to this entry.</param>
+    /// <param name="logFormatter">The custom log entry formatter.</param>
+    private object GetFormattedLogEntry(LogLevel logLevel, DateTime timestamp, object message,
+        Exception exception, ILogFormatter logFormatter)
+    {
+        if (logFormatter is null)
+            return null;
+
+        var logEntry = new LogEntry
+        {
+            Timestamp = timestamp,
+            Level = logLevel,
+            Service = Service,
+            Name = _name,
+            Message = message,
+            Exception = exception,
+            SamplingRate = CurrentConfig.SamplingRate,
+        };
+
+        var extraKeys = new Dictionary<string, object>();
+
+        // Add Custom Keys
+        foreach (var (key, value) in Logger.GetAllKeys())
+        {
+            switch (key)
+            {
+                case LoggingConstants.KeyColdStart:
+                    logEntry.ColdStart = (bool)value;
+                    break;
+                case LoggingConstants.KeyXRayTraceId:
+                    logEntry.XRayTraceId = value as string;
+                    break;
+                case LoggingConstants.KeyCorrelationId:
+                    logEntry.CorrelationId = value as string;
+                    break;
+                default:
```

Comment:
In here we are changing the premise that what is logged should be exactly what is returned from the formatter.

Reply:
We don't change the object returning from formatter and we log it as is. This is not for the object returned from the formatter, it's for creating an LogEntry to pass it to the formatter.